packages/flutter/lib/src/widgets/animated_scroll_view.dart

```diff
@@ -682,18 +682,35 @@ typedef AnimatedListRemovedItemBuilder = Widget Function(BuildContext context, A
 ///   instead of removing them.
 typedef AnimatedRemovedItemBuilder = Widget Function(BuildContext context, Animation<double> animation);
 
+/// Signature for the builder callback used in [AnimatedListState.insertItem] and
+/// [AnimatedGridState.insertItem] to animate their children when they are
+/// getting inserted.
+///
+/// The `context` argument is the build context where the widget will be
+/// created, and the `animation` is an [Animation] that should be used to
+/// animate an insert transition for the widget that is built.
+///
+/// See also:
+///
+/// * [AnimatedItemBuilder], a builder that is for adding items with animations.
+typedef AnimatedInsertedItemBuilder = Widget Function(BuildContext context, Animation<double> animation);
```

Comment:
I think AnimatedItemBuilder is used for inserting, I think I am missing why this is needed. What happens if both itemBuilder and this are set?

Reply:
It will prefer the AnimatedInsertedItemBuilder over the AnimatedItemBuilder to allow a custom insert animation. If the AnimatedInsertedItemBuilder is not set it will use the AnimatedItemBuilder.

Reply:
I am not sure that this change is necessary then, it sounds like this is just another way to do the same that can be done with AnimatedItemBuilder? Increasing the number of options can complicate an API. Is there a scenario where AnimatedItemBuilder does not work?

Reply:
`AnimatedItemBuilder` would also work in every case, but we need to consider the following options: 

- Create a new type `AnimatedInsertedItemBuilder` similar to `AnimatedRemovedItemBuilder` 
  - \+ align the *insert animation flow* to the *remove animation flow*
  - \+ no redundant information about the index
  - \- new type required
- Refactor `AnimatedRemovedItemBuilder` to be used for inserting and removing
  - \+ align the *insert animation flow* to the *remove animation flow*
  - \+ no redundant information about the index
  - \- refactoring
- Use the `AnimatedItemBuilder` type for insert animations
  - \+ no new type required
  - \- information about the index is redundant
  - \- not aligned with the *remove animation flow* 
- Remove `AnimatedRemovedItemBuilder` and consistently use `AnimatedItemBuilder` for every operation
  - \+ builders can easily be reused because of the same signature
  - \+ simplify the API
  - \- index information is redundant
  - \- **breaking change**

Reply:
> AnimatedItemBuilder would also work in every case

If it works, why are any of the listed changes above necessary? I don't know if I am understanding what the issue is that is being resolved, can you explain further? It is possible to animate items being inserted, and items being removed. There is an example in the API docs that illustrates this, where you can have different animations/builders for inserting and removing if you so desired: https://api.flutter.dev/flutter/widgets/AnimatedList-class.html#widgets.AnimatedList.1

Can you also elaborate on where you mentioned aspects of the API `aligning`? I am not sure what this means, does it mean aligning in so far as the name of the typedef?

Reply:
Sorry for the confusion, let me elaborate. When inserting an item we need to now which item at which index we are inserting thus making the information about the index in the builder unnecessary. Also if we don't use the index in the builder it is more similar to the remove item builder (thats what I meant with aligned)